cmd/nerdctl/compose.go

```diff
@@ -27,62 +27,91 @@ import (
 	"github.com/containerd/nerdctl/pkg/imgutil"
 	"github.com/containerd/nerdctl/pkg/netutil"
 	"github.com/pkg/errors"
-	"github.com/urfave/cli/v2"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
 )
 
-var composeCommand = &cli.Command{
-	Name:  "compose",
-	Usage: "Compose",
-	Subcommands: []*cli.Command{
-		composeUpCommand,
-		composeLogsCommand,
-		composeBuildCommand,
-		composeDownCommand,
-	},
-
-	Flags: []cli.Flag{
-		&cli.StringFlag{
-			Name:    "file",
-			Aliases: []string{"f"},
-			Usage:   "Specify an alternate compose file",
-		},
-		&cli.StringFlag{
-			Name:  "project-directory",
-			Usage: "Specify an alternate working directory",
-		},
-		&cli.StringFlag{
-			Name:    "project-name",
-			Aliases: []string{"p"},
-			Usage:   "Specify an alternate project name",
-		},
-		&cli.StringFlag{
-			Name:  "env-file",
-			Usage: "Specify an alternate environment file",
-		},
-	},
+func newComposeCommand() *cobra.Command {
+	var composeCommand = &cobra.Command{
+		Use:   "compose",
+		Short: "Compose",
+	}
+	composeCommand.PersistentFlags().StringP("file", "f", "", "Specify an alternate compose file")
+	composeCommand.PersistentFlags().String("project-directory", "", "Specify an alternate working directory")
+	composeCommand.PersistentFlags().StringP("project-name", "p", "", "Specify an alternate project name")
+	composeCommand.PersistentFlags().String("env-file", "", "Specify an alternate environment file")
+
+	composeCommand.AddCommand(
+		newComposeUpCommand(),
+		newComposeLogsCommand(),
+		newComposeBuildCommand(),
+		newComposeDownCommand(),
+	)
+
+	return composeCommand
 }
 
-func getComposer(clicontext *cli.Context, client *containerd.Client) (*composer.Composer, error) {
-	nerdctlCmd, nerdctlArgs := globalFlags(clicontext)
+func getComposer(cmd *cobra.Command, client *containerd.Client) (*composer.Composer, error) {
+	nerdctlCmd, nerdctlArgs := globalFlags(cmd)
+	projectDirectory, err := cmd.Flags().GetString("project-directory")
+	if err != nil {
+		return nil, err
+	}
+	envFile, err := cmd.Flags().GetString("env-file")
+	if err != nil {
+		return nil, err
+	}
+	projectName, err := cmd.Flags().GetString("project-name")
+	if err != nil {
+		return nil, err
+	}
+	debugFull, err := cmd.Flags().GetBool("debug-full")
+	if err != nil {
+		return nil, err
+	}
+	file, err := cmd.Flags().GetString("file")
+	if err != nil {
+		return nil, err
+	}
+	insecure, err := cmd.Flags().GetBool("insecure-registry")
+	if err != nil {
+		return nil, err
+	}
+	cniPath, err := cmd.Flags().GetString("cni-path")
+	if err != nil {
+		return nil, err
+	}
+	cniNetconfpath, err := cmd.Flags().GetString("cni-netconfpath")
+	if err != nil {
+		return nil, err
+	}
+	snapshotter := viper.GetString("snapshotter")
```

Comment:
why viper?

Reply:
If you want get value from envvar, and fallback to pflag, you have to use viper.

Reply:
but the envvar for snpahotter is `CONTAINERD_SNAPSHOTTER` ,  I assume you should use `getFlagOrEnv` for the snapshotter